osbs/tekton.py

```diff
@@ -590,180 +355,357 @@ def log_and_sleep():
-                logger.debug("is %s in %s?", repr(obj_status_lower), states)
-                if obj_status_lower in states:
-                    logger.debug("Yes, build is in the state I'm waiting for.")
-                    return obj
-                else:
-                    logger.debug("No, build is not in the state I'm "
-                                 "waiting for.")
-            else:
-                logger.info("The build %r isn't me %r", obj_name, build_id)
-
-        # I'm not sure how we can end up here since there are two possible scenarios:
-        #   1. our object was found and we are returning in the loop
-        #   2. our object was not found and we keep waiting (in the loop)
-        # Therefore, let's raise here
-        logger.warning("build '%s' was not found during wait", build_id)
-        raise OsbsWatchBuildNotFound("build '%s' was not found and response stream ended" %
-                                     build_id)
-
-    def wait_for_build_to_finish(self, build_id):
-        for retry in range(WAIT_RETRY):
-            try:
-                build_response = self.wait(build_id, BUILD_FINISHED_STATES)
-                return build_response
-            except OsbsWatchBuildNotFound:
-                # this is woraround for https://github.com/openshift/origin/issues/2348
-                logger.warning("I'm going to wait again. Retry #%d.", retry)
-                continue
-        raise OsbsException("Failed to wait for a build: %s" % build_id)
+                # (Unknown, Started), (Unknown, PipelineRunCancelled)
+                logger.debug("Waiting for pipeline run, current status %s, reason %s", status, reason)
 
-    def wait_for_build_to_get_scheduled(self, build_id):
-        for _ in range(WAIT_RETRY):
+    def wait_for_taskruns(self):
+        for changetype, pipeline_run in self.os.watch_resource(self.api_path, self.api_version, resource_type='pipelineruns', resource_name=self.pipeline_run_name):
             try:
-                build_response = self.wait(build_id, BUILD_FINISHED_STATES + BUILD_RUNNING_STATES)
-                return build_response
-            except OsbsWatchBuildNotFound:
+                status = pipeline_run['status']['conditions'][0]['status']
+                reason = pipeline_run['status']['conditions'][0]['reason']
+                task_runs = pipeline_run['status']['taskRuns']
+            except KeyError:
+                logger.error("pipeline run does not have any task runs")
                 continue
-        raise OsbsException('Failed to schedule a build in {} attempts: {}'.format(WAIT_RETRY,
-                                                                                   build_id))
+            # pipeline run finished succesfully or failed
+            if status in ['True', 'False']:
+                yield task_runs
+            elif status == 'Unknown' and reason == 'Running':
+                yield task_runs
```

Comment:
I'm having a hard time understanding what the method does. It yields a list of task runs every time the status changes, that much I understand, but what does that list actually mean? Is it a list of task runs currently in progress? A list of all the task runs that were created up until that point? Something else?

Would it make sense to de-duplicate the taskruns here and flatten the result from `Iterator[List[str]]` to `Iterator[str]`?

And is it necessary to check the status+reason?

Reply:
+1 can you add explanation docstring to this method please?

Reply:
I added explanation to docstring, and cleared up that method as per adam's suggestions.